features/268-next-cf/api.md

```diff
@@ -0,0 +1,627 @@
+
+Skycli may have command to set the current version of a function to a specific version.
+
+Update: [CF versioning](./versioning.md)
+
+# Function configuration
+
+Function configuration may be placed in the file for other skycli configuration.
+
+Example:
+
+```yaml
+cf:
+  function1:
+    type: http-handler
+    path: /hello-world
+    env: node
+    src: js/hellow-world
+    secrets:
+      - DATABASE_URL
+    permission:
+      - name: key_required
+      - name: user_required
+```
+
+- `cf` indicates the dictionary is the cf configuration in the file
+- `function1` is the name of function, function deployed with the same name will considered as the same function with different version
+- `type` is the type of the function
+  - `http-handler`
+    - a single function that accepts http requests from an exact path
+    - anytime may be scaled down to zero
+  - `http-service`
+    - an http server that accepts http requests from the path
+    - may be long running
+  - `function`:
+    - a single function (in a program) that read input from stdin and write output to stdout
+    - anytime may be scaled down to zero
+- `env` indicates the environment that the code would be built with, e.g.
+  - `docker` is the base environment that the Dockerfile would be deployed directly, or would be wrapped by another minimal Dockerfile
+  - environment of other languages (e.g. `node`, `python`, `golang`) would accept source files from the developers and build with a Dockerfile provided by us
+- `src` may specify where to find Dockerfile or source code files
+- `secrets` list what secrets would be passed to the functions, developer need to set secrets with skycli
+- `permission` describes how the request can access the function
+  - `key_required`, if true, need API key or master key
+  - `user_required`, if true, need access token of a verified and not disabled user
+  - `admin_required`, if true, need the user have admin role
+  - `role_required` with `roles` tells the function require specified roles
+
+# Serving content
+
+## Server side rendering
+
+`http-service` would fulfill server side rendering because it works on http request and response, developers can
+- get the path of the request after the gateway rewrite
+- set the corresponding http headers and write content to the response body
+
+## Static assets
+
+In skygear v1 cloud, requests with path `/static/*` will go to s3 endpoint with proxy pass. This implies that
+- these requests will be served with the same domain of the api server
+- the content data will always pass through the skygear cluster
+
+Next cloud will serve the content with another domain, e.g. `myapp.skygear.io` for static assets v.s. `myapp.api.skygear.io` for functions and gear api, because
```

Comment:
Just want to ask if we will reconsider this. It seems if we could have the versioning on this and serve static assets via same domains, the benefits are quite great:

1. CORS or some security troubles will be gone.
1. Developers develop SPA/PWA could use Skygear platform as a total solutions for deployment and versioning besides DB.

Reply:
Yes, I think so, since our original concern was about the CDN, and we figured out there should be a feasible way to put CDN in front of skygear without special requirement on domain setting.

Reply:
Looking forward to the update.